### DISCUSSION TITLE
Question about "slot missing" error.

### DISCUSSION THREAD

**Discussion by aykaramba** [fukamachi/mito#164] [General]
As short a description as possible:

1) Database: SQLite

2) Work done to upgrade my knowledge:

* I have completed various SQL tutorials online.

* I use DB Browser for Sqlite to see what is going on in the database.  I also write sql queries in this tool first to verify that I have the syntax correct before trying to create sxql equivalent queries.

* I have gone through the Mito and SxQL examples on github and on the net.

* I have completed Vindarels CLOS tutorials on Udemy once again in order to refresh my self on CLOS.

3) I am getting the hang of how and where to use Mito and SxQL.  It is still confusing because Mito seems to have some query capabilities of SxQL suggesting that at some point in time it will gain more sophisticated query capabilities.  

Right now I believe I understand that we use SxQL for the fine grained queries in a database and Mito for generating CLOS objects that we can inspect.

4) I created a simple SQLite db with Mito called "docs" and created a single column: "document".  Mito created a database with the following columns (which is great):  id, document, created_at, updated_at.

5) I can retrieve lists of raw data (strings) using (mito:retrieve-by-sql ... ) but then I have to car / cdr through the lists to grab the data.  This doesn't feel right.

I thought I could just use mito to generate a CLOS obejct and inspect the slot value like this:

a) We can get a list of objects using this:  (mito:select-dao 'docs)

This returns a list of CLOS objects, all of which we can inspect in SLY and see that they have the slots and values.

b) Now, let's say that we want a single CLOS object where id = 37:  (mito:select-dao 'docs (sxql:where (:= :id 37)))

This returns a list of CLOS objects (in our case 1 object): (#<DOCS {1004AF7153}>)

... and if we inspect it ... 

```
#<DOCS {1004AF7153}>
--------------------
Class: #<DAO-TABLE-CLASS TOKYO::DOCS>
--------------------
 Group slots by inheritance [ ]
 Sort slots alphabetically  [X]

All Slots:
[ ]  CREATED-AT = @2024-11-30T12:57:39.749641-08:00
[ ]  DOCUMENT   = "blah blah blah"
[ ]  ID         = 28
[ ]  SYNCED     = T
[ ]  UPDATED-AT = @2024-11-30T12:57:39.749641-08:00

[set value]  [make unbound]
```

Here we see 5 slots what we ought to be able to inspect for slot-value.

c) Since it is a list of CLOS objects, we should be able to access slot values like this:

(slot-value (car (mito:select-dao 'docs (sxql:where (:= :id 37)))) 'document)

... and we get the content of the document column for :id 37 which is "blah blah blah".   However, the following don't work and the environment says the slots don't exist for the very same object:

(slot-value (car (mito:select-dao 'docs (sxql:where (:= :id 37)))) 'id)

(slot-value (car (mito:select-dao 'docs (sxql:where (:= :id 37)))) 'created-at)

(slot-value (car (mito:select-dao 'docs (sxql:where (:= :id 37)))) 'updated-at)

In the CLOS tutorials that I took online, if a CLOS object had slots and they have an :accessor we could query the data for the slot.  The SLY inspector shows that the CLOS object has slots and it shows the data in the slots.  How is it that I am not able to do (slot-vale .... 'created-at) on an object?  Does Mito not create an accessor for that slot?

What might I be doing wrong?  Any suggestions or tips?

**Comment by vindarel** [comment]
>  using (mito:retrieve-by-sql ... ) but then I have to car / cdr through the lists to grab the data. This doesn't feel right.

correct, this function can be used in the last resort for special SQL queries that don't fit with another method.

> (mito:select-dao 'docs (sxql:where (:= :id 37)))

you can shorten this to

    (mito:find-dao 'docs :id 37)

(see Mito's README and the Cookbook)

>  How is it that I am not able to do (slot-vale .... 'created-at) on an object?

I didn't try with an example but this rings a bell (and like, very loudly): beware of the packages. What package are you in and from what package does the `'created-at` symbol pertain? 

Remind that `cl-user:created-at` is NOT the same as `foobarbaz:created-at` or `mito:created-at`. 

It's an easy trap to fall in where I've been a few times.

We could check this is the mistake you made if you show us the full error message.

**Comment by aykaramba** [comment]
Thanks for you reply!  Greatly appreciate any insight.

So:

1) My package is called mypackage.

2) I start up emacs, start up sly and (ql:quickload :mypackage)

3) In sly I move into the package with (in-package :mypackage)

4) We can now do: (mito:find-dao 'docs :id 37) and get an object back: #<DOCS {1006F16FD3}>

5) And we can take a peek at the document it self with: (slot-value (mito:find-dao 'docs :id 37) 'document) and we get the correct data: 'blah blah blah'

6) The table has 4 columns: id, document, created_at, updated_at.  Can we query the created-at slot?  Let's try: (slot-value (mito:find-dao 'docs :id 37) 'updated-at)

We get the following error: 

```
When attempting to read the slot's value (slot-value), the slot
CREATED-AT is missing from the object #<DOCS {100427CCA3}>.
It has a slot MITO.DAO.MIXIN::CREATED-AT, while mypackage::CREATED-AT is requested.
   [Condition of type SB-PCL::MISSING-SLOT]

Restarts:
 0: [RETRY] Retry SLY mREPL evaluation request.
 1: [*ABORT] Return to SLY's top level.
 2: [ABORT] abort thread (#<THREAD "sly-channel-1-mrepl-remote-1" RUNNING {10015180A3}>)

Backtrace:
 0: ((:METHOD SLOT-MISSING (T T T T)) #<DAO-TABLE-CLASS mypackage::DOCS> #<DOCS {100427CCA3}> CREATED-AT SLOT-VALUE NIL) [fast-method]
      Locals:
        CLASS = #<DAO-TABLE-CLASS mypackage::DOCS>
        SB-FORMAT::FORMAT-ARG1 = NIL
        SB-KERNEL:INSTANCE = #<DOCS {100427CCA3}>
        SB-PCL::OPERATION = SLOT-VALUE
        SB-PCL::SLOT-NAME = CREATED-AT
 1: (SLOT-VALUE #<DOCS {100427CCA3}> CREATED-AT)
      Locals:
        SB-PCL::OBJECT = #<DOCS {100427CCA3}>
        SB-PCL::SLOT-NAME = CREATED-AT
 2: (SB-INT:SIMPLE-EVAL-IN-LEXENV (SLOT-VALUE (FIND-DAO (QUOTE DOCS) :ID 37) (QUOTE CREATED-AT)) #<NULL-LEXENV>)
      Locals:
        SB-KERNEL:LEXENV = #<NULL-LEXENV>
        SB-IMPL::ORIGINAL-EXP = (SLOT-VALUE (FIND-DAO 'DOCS :ID 37) 'CREATED-AT)
 3: (EVAL (SLOT-VALUE (FIND-DAO (QUOTE DOCS) :ID 37) (QUOTE CREATED-AT)))
      Locals:
        SB-IMPL::ORIGINAL-EXP = (SLOT-VALUE (FIND-DAO 'DOCS :ID 37) 'CREATED-AT)
 4: ((LAMBDA NIL :IN SLYNK-MREPL::MREPL-EVAL-1))
      Locals:
        SLYNK-MREPL::FORM = (SLOT-VALUE (FIND-DAO 'DOCS :ID 37) 'CREATED-AT)
        SLYNK-MREPL::IN = #<SB-IMPL::STRING-INPUT-STREAM {7F96A7DADE73}>
 5: (SLYNK::CALL-WITH-RETRY-RESTART "Retry SLY mREPL evaluation request." #<FUNCTION (LAMBDA NIL :IN SLYNK-MREPL::MREPL-EVAL-1) {100427A33B}>)
      Locals:
        MSG = "Retry SLY mREPL evaluation request."
        THUNK = #<FUNCTION (LAMBDA () :IN SLYNK-MREPL::MREPL-EVAL-1) {100427A33B}>
 6: ((LAMBDA NIL :IN SLYNK-MREPL::MREPL-EVAL-1))
      [No Locals]
 7: ((LAMBDA NIL :IN SLYNK::CALL-WITH-LISTENER))
      Locals:
        SB-C::OBJECT = #<SLYNK-MREPL::MREPL mrepl-1-1>
        SLYNK::SAVING = NIL
 8: (SLYNK::CALL-WITH-BINDINGS ((*PACKAGE* . #<PACKAGE "mypackage">) (*DEFAULT-PATHNAME-DEFAULTS* . #P"/home/user/common-lisp/mypackage/") (* . #1="you called?") (** . "blah blah blah"..
      Locals:
        ALIST = ((*PACKAGE* . #<PACKAGE "mypackage">) (*DEFAULT-PATHNAME-DEFAULTS* . #P"/home/user/common-lisp/mypackage/") (* . "you called?") (** . "blah blah blah") ..)
        FUN = #<FUNCTION (LAMBDA () :IN SLYNK::CALL-WITH-LISTENER) {100427A30B}>
 9: (SLYNK-MREPL::MREPL-EVAL-1 #<SLYNK-MREPL::MREPL mrepl-1-1> "(slot-value (mito:find-dao 'docs :id 37) 'created-at)")
      Locals:
        REPL = #<SLYNK-MREPL::MREPL mrepl-1-1>
        STRING = "(slot-value (mito:find-dao 'docs :id 37) 'created-at)"
10: (SLYNK-MREPL::MREPL-EVAL #<SLYNK-MREPL::MREPL mrepl-1-1> "(slot-value (mito:find-dao 'docs :id 37) 'created-at)")
      Locals:
        ABORTED = #<SB-PCL::MISSING-SLOT CREATED-AT {100427CE93}>
        REPL = #<SLYNK-MREPL::MREPL mrepl-1-1>
        RESULTS = NIL
        STRING = "(slot-value (mito:find-dao 'docs :id 37) 'created-at)"
11: (SLYNK:PROCESS-REQUESTS NIL)
      Locals:
        TIMEOUT = NIL
      Catch-tags:
        SLYNK::STOP-PROCESSING
12: ((LAMBDA NIL :IN SLYNK::SPAWN-CHANNEL-THREAD))
      Locals:
        SLYNK::CHANNEL = #<SLYNK-MREPL::MREPL mrepl-1-1>
13: ((LAMBDA NIL :IN SLYNK::SPAWN-CHANNEL-THREAD))
      Locals:
        SLYNK::CHANNEL = #<SLYNK-MREPL::MREPL mrepl-1-1>
        SLYNK::CONNECTION = #<SLYNK::MULTITHREADED-CONNECTION {1003937D63}>
14: (SLYNK-SBCL::CALL-WITH-BREAK-HOOK #<FUNCTION SLYNK:SLYNK-DEBUGGER-HOOK> #<FUNCTION (LAMBDA NIL :IN SLYNK::SPAWN-CHANNEL-THREAD) {100171000B}>)
      Locals:
        CONTINUATION = #<FUNCTION (LAMBDA () :IN SLYNK::SPAWN-CHANNEL-THREAD) {100171000B}>
        HOOK = #<FUNCTION SLYNK:SLYNK-DEBUGGER-HOOK>
15: ((FLET SLYNK-BACKEND:CALL-WITH-DEBUGGER-HOOK :IN "/home/user/.emacs.d/elpa/sly-20230523.1254/slynk/backend/sbcl.lisp") #<FUNCTION SLYNK:SLYNK-DEBUGGER-HOOK> #<FUNCTION (LAMBDA NIL :IN SLYNK::SPAWN-..
      Locals:
        SLYNK-SBCL::FUN = #<FUNCTION (LAMBDA () :IN SLYNK::SPAWN-CHANNEL-THREAD) {100171000B}>
        SLYNK-SBCL::HOOK = #<FUNCTION SLYNK:SLYNK-DEBUGGER-HOOK>
16: ((LAMBDA NIL :IN SLYNK::CALL-WITH-LISTENER))
      Locals:
        SB-C::OBJECT = #<SLYNK-MREPL::MREPL mrepl-1-1>
        SLYNK::SAVING = NIL
17: (SLYNK::CALL-WITH-BINDINGS ((*PACKAGE* . #<PACKAGE "mypackage">) (*DEFAULT-PATHNAME-DEFAULTS* . #P"/home/user/common-lisp/mypackage/") (* . #1="you called?") (** . "blah blah blah"..
      Locals:
        ALIST = ((*PACKAGE* . #<PACKAGE "mypackage">) (*DEFAULT-PATHNAME-DEFAULTS* . #P"/home/user/common-lisp/mypackage/") (* . "you called?") (** . "blah blah blah") ..)
        FUN = #<FUNCTION (LAMBDA () :IN SLYNK::CALL-WITH-LISTENER) {100171004B}>
18: ((LAMBDA NIL :IN SLYNK::SPAWN-CHANNEL-THREAD))
      Locals:
        SLYNK::CONNECTION = #<SLYNK::MULTITHREADED-CONNECTION {1003937D63}>
19: ((FLET SB-UNIX::BODY :IN SB-THREAD::RUN))
      [No Locals]
      Catch-tags:
        SB-THREAD::%RETURN-FROM-THREAD
20: ((FLET "WITHOUT-INTERRUPTS-BODY-11" :IN SB-THREAD::RUN))
      [No Locals]
21: ((FLET SB-UNIX::BODY :IN SB-THREAD::RUN))
      [No Locals]
      Catch-tags:
        SB-THREAD::%ABORT-THREAD
        SB-IMPL::%END-OF-THE-WORLD
        SB-INT:TOPLEVEL-CATCHER
22: ((FLET "WITHOUT-INTERRUPTS-BODY-4" :IN SB-THREAD::RUN))
      [No Locals]
23: (SB-THREAD::RUN)
      [No Locals]
24: ("foreign function: call_into_lisp_")
      [No Locals]
25: ("foreign function: funcall1")
      [No Locals]
```

7) The slots exist for the object:   #<DOCS {1007093543}>

```
--------------------
Class: #<DAO-TABLE-CLASS mypackage::DOCS>
--------------------
 Group slots by inheritance [ ]
 Sort slots alphabetically  [X]

All Slots:
[ ]  CREATED-AT = @2024-11-30T12:57:48.954586-08:00
[ ]  DOCUMENT   = "blah blah blah"
[ ]  ID         = 37
[ ]  SYNCED     = T
[ ]  UPDATED-AT = @2024-11-30T12:57:48.954586-08:00

[set value]  [make unbound]
```

This still feels like I am doing something wrong, but I cannot figure it out.  Thoughts?

**Reply by vindarel** [reply]
you're nearly there. Read carefully:

>  the slot CREATED-AT is missing from the object #<DOCS {100427CCA3}>. It has a slot MITO.DAO.MIXIN::CREATED-AT

the trick is that created-at was added by Mito. The error message shows us it is a symbol living in the mito.dao.mixin package. So:

```lisp
(slot-value (mito:find-dao 'docs :id 37) 'mito.dao.mixin::created-at)
```

BUT

there is a built-in Mito method for this, use:

```lisp
(mito:object-created-at **)
```

**Reply by vindarel** [reply]
How could you know? You saw this representation and you thought you could access the created-at slot:

```

All Slots:
[ ]  CREATED-AT = @2024-11-30T12:57:48.954586-08:00
…
```

but place the cursor on "created-at", press Enter to see the description of the slot, you'll see:

```
#<SB-MOP:STANDARD-EFFECTIVE-SLOT-DEFINITION {100E2B5A53}>
--------------------
Name: MITO.DAO.MIXIN::CREATED-AT
Init args: (:CREATED-AT)
Init form: #<unspecified>
Init function: NIL
--------------------

…
```

where you find out its complete name, `mito.dao.mixin::created-at`.

**Reply by vindarel** [reply]
And again, how could you know? Another method, with a bit of introspection.

```lisp
(ql:quickload "closer-mop")

;; (example of mine)
(mopp:class-slots (find-class 'shelf))

;; =>
(#<SB-MOP:STANDARD-EFFECTIVE-SLOT-DEFINITION MITO.DAO.MIXIN::CREATED-AT>
 #<SB-MOP:STANDARD-EFFECTIVE-SLOT-DEFINITION MITO.DAO.MIXIN::UPDATED-AT>
 #<SB-MOP:STANDARD-EFFECTIVE-SLOT-DEFINITION MITO.DAO.MIXIN::SYNCED>
 #<SB-MOP:STANDARD-EFFECTIVE-SLOT-DEFINITION MITO.DAO.MIXIN::ID>
 #<SB-MOP:STANDARD-EFFECTIVE-SLOT-DEFINITION COSMO-ADMIN-DEMO::NAME>)
```

In the results we clearly see that there are definitions that are in the mito.dao.mixin package, only the NAME slot is in my package (cosmo-admin-demo).

**Comment by aykaramba** [comment]
Ohhhhh.  I needed to read and think a little bit more.  The problem is between my two ears.

Thank you very kindly for this tutorial.  I have slowly learned how to read the information provided by the system, this helps me drill down even more.

Also, really appreciate the deep dive at the end on "closer-mop".    That is incredibly useful.  As I am learning Common Lisp, I feel like the entire ecosystem (Slime et al.) is more of a framework than just a programming language.

Thank you very kindly.
